synapse/third_party_rules/access_rules.py

```diff
@@ -593,7 +640,7 @@ def _on_join_rule_change(self, event: EventBase, rule: str) -> bool:
             Whether the change is allowed.
         """
         if event.content.get("join_rule") == JoinRules.PUBLIC:
-            return rule == AccessRules.RESTRICTED
+            return rule != AccessRules.DIRECT
 
```

Comment:
This is testing that a room that's direct can't be public, do we need to care about join rules that aren't public nor invite?

Reply:
Actually https://matrix.org/docs/spec/client_server/latest#m-room-join-rules says other values are reserved values and nothing more; however one of them is `knock`, which should be implemented soon, so I think we should cover it here.

Reply:
This can probably be handled in another PR, just wanted to mention it so we have a record of it.